src/front/wgsl/parse/ast.rs

```diff
@@ -0,0 +1,284 @@
+use crate::{front::wgsl::text::Text, Span};
+
+#[derive(Clone, Debug, Default)]
+pub struct TranslationUnit {
```

Comment:
This AST is very close to the grammar, but I don't think it's worth building all this intermediate structure. We should just touch the data once, lower it as far as we can immediately, and be done with it. The architecture here leads to odd bits like the check for generic parameters on the identifier in `Resolver::expr`; it seems like the code could be structured so that situations like that (and others) simply never arise.

Naga tries hard to gather everything into arrays (the `Arena` types, for example) and avoid heap allocation because such code runs very well on modern processors. As I say, I haven't profiled, but my guess is that you won't be able to get within 2x or 3x of the current Naga front end's run time without addressing this.